### PR TITLE
feat(core, apps): 'PacketDataProvider' interface added and implemented

### PR DESCRIPTION
**Pull request by srdtrk**
## Description

This PR reduces the diff when reviewing the adr8 PR https://github.com/cosmos/ibc-go/pull/3939.

`PacketDataProvider` defines an optional interfaces for retrieving custom packet data stored on behalf of another application.

closes: #XXXX

### Commit Message / Changelog Entry

```text
feat(core, apps): 'PacketDataProvider' interface added and implemented
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] ~Linked to Github issue with discussion and accepted design OR link to spec that describes this work.~
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [x] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/exported/packet.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	// Note 2: The return value should account for the success of the underlying IBC application or middleware. Thus the  `acknowledgement.Success` is representative of the entire IBC stack's success when receiving a packet. The individual success of each acknowledgement associated with an IBC application or middleware must be determined by obtaining the actual acknowledgement type after decoding the acknowledgement bytes.
+	//
+	// See https://github.com/cosmos/ibc-go/blob/v7.0.0/docs/ibc/apps.md for further explanations.
+	Success() bool
+	Acknowledgement() []byte
+}
+
+// PacketDataProvider defines an optional interfaces for retrieving custom packet data stored on behalf of another application.
+// An existing problem in the IBC middleware design is the inability for a middleware to define its own packet data type and insert packet sender provided information.
+// A short term hack was introduced into several packet data's to utilize a memo to carry this information on behalf of another application.
```
Comment (crodriguezvega) [inline modules/core/exported/packet.go:38]: :)

```suggestion
// A short term solution was introduced into several application's packet data to utilize a memo field to carry this information on behalf of another application.
```

File: modules/apps/27-interchain-accounts/types/packet_test.go [hunk 2]
```diff
@@ -82,3 +85,70 @@ func (suite *TypesTestSuite) TestValidateBasic() {
...
 	}
 }
+
+func (suite *TypesTestSuite) TestPacketDataProvider() {
+	expCallbackAddr := ibctesting.TestAccAddress
+
+	testCases := []struct {
+		name              string
+		packetData        types.InterchainAccountPacketData
+		expAdditionalData map[string]interface{}
```
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/types/packet_test.go:95]: Maybe `expCustomData` or `expMemoData`?
```suggestion
		expCustomData map[string]interface{}
```

**Review by crodriguezvega — APPROVED** [review]

**Review by colin-axner — APPROVED** [review]
LGTM! Will take another quick peak after the cast removal change is added

File: modules/core/exported/packet.go (hunk 1 above)
Comment (srdtrk) [reply modules/core/exported/packet.go]: done

File: modules/apps/27-interchain-accounts/types/packet_test.go (hunk 2 above)
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/types/packet_test.go]: done

File: modules/core/exported/packet.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+// This interfaces standardizes that behaviour. Upon realization of the ability for middleware's to define their own packet data types, this interface will be deprecated and removed with time.
+type PacketDataProvider interface {
+	// GetCustomPacketData returns the packet data held on behalf of another application.
+	// The name the information is stored under should be provided as the key
+	// If no custom packet data exists for the key, nil is returned.
+	GetCustomPacketData(key string) interface{}
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/core/exported/packet.go]: done

File: modules/apps/transfer/types/packet.go [hunk 4]
```diff
@@ -64,3 +68,25 @@ func (ftpd FungibleTokenPacketData) ValidateBasic() error {
...
+		return nil
+	}
+
+	jsonObject := make(map[string]interface{})
+	err := json.Unmarshal([]byte(ftpd.Memo), &jsonObject)
+	if err != nil {
+		return nil
+	}
+
+	memoData, ok := jsonObject[key].(map[string]interface{})
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/types/packet.go]: updated

File: modules/apps/transfer/types/packet.go [hunk 5]
```diff
@@ -64,3 +68,25 @@ func (ftpd FungibleTokenPacketData) ValidateBasic() error {
 func (ftpd FungibleTokenPacketData) GetBytes() []byte {
 	return sdk.MustSortJSON(mustProtoMarshalJSON(&ftpd))
 }
+
+// GetCustomPacketData returns a json object from the memo as `map[string]interface{}` so that
+// it can be interpreted as a json object with keys.
+// If the key is missing or the memo is not properly formatted, then nil is returned.
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/types/packet.go]: updated

File: modules/apps/transfer/types/packet_test.go [hunk 6]
```diff
@@ -43,3 +44,84 @@ func TestFungibleTokenPacketDataValidateBasic(t *testing.T) {
 		}
 	}
 }
+
+func (suite *TypesTestSuite) TestPacketDataProvider() {
+	testCases := []struct {
+		name              string
+		packetData        types.FungibleTokenPacketData
+		expAdditionalData map[string]interface{}
+		expPacketSender   string
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/types/packet_test.go]: removed

File: modules/apps/transfer/types/packet_test.go [hunk 7]
```diff
@@ -43,3 +44,84 @@ func TestFungibleTokenPacketDataValidateBasic(t *testing.T) {
 		}
 	}
 }
+
+func (suite *TypesTestSuite) TestPacketDataProvider() {
+	testCases := []struct {
+		name              string
+		packetData        types.FungibleTokenPacketData
+		expAdditionalData map[string]interface{}
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/types/packet_test.go]: done

**Review by damiannolan — APPROVED** [review]

File: modules/apps/transfer/types/packet_test.go [hunk 8]
```diff
@@ -43,3 +44,77 @@ func TestFungibleTokenPacketDataValidateBasic(t *testing.T) {
 		}
 	}
 }
+
+func (suite *TypesTestSuite) TestPacketDataProvider() {
+	testCases := []struct {
+		name              string
+		packetData        types.FungibleTokenPacketData
+		expAdditionalData interface{}
```
Comment (colin-axner) [inline modules/apps/transfer/types/packet_test.go:52]: super super nit: `expAdditionalData` -> `expCustomData`

**Review by colin-axner — APPROVED** [review]
Beautiful! Thank you! Only left a comment on a nit because of naming consistency

File: modules/apps/transfer/types/packet_test.go (hunk 8 above)
Comment (srdtrk) [reply modules/apps/transfer/types/packet_test.go]: done

**Pull request merged**